### PR TITLE
8185887: TableRowSkinBase fails to correctly virtualize cells in horizontal direction

### PR DESCRIPTION
**Pull request by yososs**
If there are many columns, the current TableView will stall scrolling. Resolving this performance issue requires column virtualization. Virtualization mode is enabled when the row height is fixed by the following method.

`tableView.setFixedCellSize(height)`

This proposal includes a fix because the current code does not correctly implement column virtualization.

The improvement of this proposal can be seen in the following test program.

``` Java
import java.util.Arrays;
import java.util.Collections;

import javafx.animation.AnimationTimer;
import javafx.application.Application;
import javafx.beans.property.SimpleStringProperty;
import javafx.collections.ObservableList;
import javafx.scene.Scene;
import javafx.scene.control.Button;
import javafx.scene.control.TableColumn;
import javafx.scene.control.TableView;
import javafx.scene.layout.BorderPane;
import javafx.scene.layout.HBox;
import javafx.stage.Stage;

public class BigTableViewTest2 extends Application {
	private static final boolean USE_WIDTH_FIXED_SIZE = false;
	private static final boolean USE_HEIGHT_FIXED_SIZE = true;
//	private static final int COL_COUNT=30;
//	private static final int COL_COUNT=300;
//	private static final int COL_COUNT=600;
	private static final int COL_COUNT = 1000;
	private static final int ROW_COUNT = 1000;

	@Override
	public void start(final Stage primaryStage) throws Exception {
		final TableView<String[]> tableView = new TableView<>();

//	    tableView.setTableMenuButtonVisible(true); //too heavy
		if (USE_HEIGHT_FIXED_SIZE) {
			tableView.setFixedCellSize(24);
		}

		final ObservableList<TableColumn<String[], ?>> columns = tableView.getColumns();
		for (int i = 0; i < COL_COUNT; i++) {
			final TableColumn<String[], String> column = new TableColumn<>("Col" + i);
			final int colIndex = i;
			column.setCellValueFactory((cell) -> new SimpleStringProperty(cell.getValue()[colIndex]));
			columns.add(column);
			if (USE_WIDTH_FIXED_SIZE) {
				column.setPrefWidth(60);
				column.setMaxWidth(60);
				column.setMinWidth(60);
			}
		}

		final Button load = new Button("load");
		load.setOnAction(e -> {
			final ObservableList<String[]> items = tableView.getItems();
			items.clear();
			for (int i = 0; i < ROW_COUNT; i++) {
				final String[] rec = new String[COL_COUNT];
				for (int j = 0; j < rec.length; j++) {
					rec[j] = i + ":" + j;
				}
				items.add(rec);
			}
		});

		final Button reverse = new Button("reverse columns");
		reverse.setOnAction(e -> {
			final TableColumn<String[], ?>[] itemsArray = columns.toArray(new TableColumn[0]);
			Collections.reverse(Arrays.asList(itemsArray));
			tableView.getColumns().clear();
			tableView.getColumns().addAll(Arrays.asList(itemsArray));
		});

		final Button hide = new Button("hide % 10");
		hide.setOnAction(e -> {
			for (int i = 0, n = columns.size(); i < n; i++) {
				if (i % 10 == 0) {
					columns.get(i).setVisible(false);
				}
			}
		});

		final BorderPane root = new BorderPane(tableView);
		root.setTop(new HBox(8, load, reverse, hide));

		final Scene scene = new Scene(root, 800, 800);
		primaryStage.setScene(scene);
		primaryStage.show();
		this.prepareTimeline(scene);
	}

	public static void main(final String[] args) {
		Application.launch(args);
	}

	private void prepareTimeline(final Scene scene) {
		new AnimationTimer() {
			@Override
			public void handle(final long now) {
				final double fps = com.sun.javafx.perf.PerformanceTracker.getSceneTracker(scene).getInstantFPS();
				((Stage) scene.getWindow()).setTitle("FPS:" + (int) fps);
			}
		}.start();
	}
}
```

``` Java
import javafx.animation.AnimationTimer;
import javafx.application.Application;
import javafx.application.Platform;
import javafx.beans.property.ReadOnlyStringWrapper;
import javafx.scene.Scene;
import javafx.scene.control.TreeItem;
import javafx.scene.control.TreeTableColumn;
import javafx.scene.control.TreeTableColumn.CellDataFeatures;
import javafx.scene.control.TreeTableView;
import javafx.stage.Stage;

public class BigTreeTableViewTest extends Application {

	private static final boolean USE_HEIGHT_FIXED_SIZE = true;
//	private static final int COL_COUNT = 900;
	private static final int COL_COUNT = 800;
//	private static final int COL_COUNT = 600;
//	private static final int COL_COUNT = 500;
//	private static final int COL_COUNT = 400;
//	private static final int COL_COUNT = 300;
//	private static final int COL_COUNT = 200;
//	private static final int COL_COUNT = 100;

	public static void main(final String[] args) {
		Application.launch(args);
	}

	@Override
	public void start(final Stage stage) {
		final TreeItem<String> root = new TreeItem<>("Root");
		final TreeTableView<String> treeTableView = new TreeTableView<>(root);
		if (USE_HEIGHT_FIXED_SIZE) {
			treeTableView.setFixedCellSize(24);
		}
		treeTableView.setPrefWidth(800);
		treeTableView.setPrefHeight(500);
		stage.setWidth(800);
		stage.setHeight(500);

		Platform.runLater(() -> {
			for (int i = 0; i < 100; i++) {
				TreeItem<String> child = this.addNodes(root);
				child = this.addNodes(child);
				child = this.addNodes(child);
				child = this.addNodes(child);
			}
		});

		final TreeTableColumn<String, String>[] cols = new TreeTableColumn[COL_COUNT + 1];
		final TreeTableColumn<String, String> column = new TreeTableColumn<>("Column");
		column.setPrefWidth(150);
		column.setCellValueFactory(
				(final CellDataFeatures<String, String> p) -> new ReadOnlyStringWrapper(p.getValue().getValue()));
		cols[0] = column;

		for (int i = 0; i < COL_COUNT; i++) {
			final TreeTableColumn<String, String> col = new TreeTableColumn<>(Integer.toString(i));
			col.setPrefWidth(60);
			col.setCellValueFactory(val -> new ReadOnlyStringWrapper(val.getValue().getValue()+":"+val.getTreeTableColumn().getText()));
			cols[i + 1] = col;
		}
		treeTableView.getColumns().addAll(cols);

		final Scene scene = new Scene(treeTableView, 800, 500);
		stage.setScene(scene);
		stage.show();
		this.prepareTimeline(scene);
	}

	private TreeItem<String> addNodes(final TreeItem<String> parent) {

		final TreeItem<String>[] childNodes = new TreeItem[20];
		for (int i = 0; i < childNodes.length; i++) {
			childNodes[i] = new TreeItem<>("N" + i);
		}
		final TreeItem<String> root = new TreeItem<>("dir");
		root.setExpanded(true);
		root.getChildren().addAll(childNodes);
		parent.setExpanded(true);
		parent.getChildren().add(root);
		return root;
	}

	private void prepareTimeline(final Scene scene) {
		new AnimationTimer() {
			@Override
			public void handle(final long now) {
				final double fps = com.sun.javafx.perf.PerformanceTracker.getSceneTracker(scene).getInstantFPS();
				((Stage) scene.getWindow()).setTitle("FPS:" + (int) fps);
			}
		}.start();
	}

}
```

---------
### Progress
- [ ] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed (2 reviews required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer), 1 [Author](https://openjdk.org/bylaws#author))

### Issue
 * [JDK-8185887](https://bugs.openjdk.org/browse/JDK-8185887): TableRowSkinBase fails to correctly virtualize cells in horizontal direction

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jfx.git pull/125/head:pull/125` \
`$ git checkout pull/125`

Update a local copy of the PR: \
`$ git checkout pull/125` \
`$ git pull https://git.openjdk.org/jfx.git pull/125/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 125`

View PR using the GUI difftool: \
`$ git pr show -t 125`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jfx/pull/125.diff">https://git.openjdk.org/jfx/pull/125.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jfx/pull/125#issuecomment-608369779)

### CODE REVIEW

**Comment by yososs** [comment]
I'm signed to OCA and already a contributor.

File: modules/javafx.controls/src/main/java/javafx/scene/control/skin/TableRowSkinBase.java [hunk 1]
```diff
@@ -429,27 +463,35 @@ public TableRowSkinBase(C control) {
...
-            } else {
-                width = snapSizeX(tableCell.prefWidth(-1)) - snapSizeX(horizontalPadding);
-
-                if (fixedCellSizeEnabled) {
-                    // we only add/remove to the scenegraph if the fixed cell
-                    // length support is enabled - otherwise we keep all
-                    // TableCells in the scenegraph
-                    getChildren().remove(tableCell);
-                }
+            }else if(fixedCellSizeEnabled && lastVisibleColumnIndex<column){
```
Comment (sghpjuikit) [inline modules/javafx.controls/src/main/java/javafx/scene/control/skin/TableRowSkinBase.java]: `}else` formatting

File: modules/javafx.controls/src/main/java/javafx/scene/control/skin/TableRowSkinBase.java [hunk 2]
```diff
@@ -339,45 +339,79 @@ public TableRowSkinBase(C control) {
...
+        final ObservableList<Node> children = getChildren();
+        if(fixedCellSizeEnabled) {
+            for (int column = 0, max = cells.size(); column < max; column++) {
+                R tableCell = cells.get(column);
+                final boolean isVisible = firstVisibleColumnIndex <= column && column <= lastVisibleColumnIndex;
+                if (isVisible ) {
+                    if(tableCell.getParent()==null){
+                        children.add(tableCell);
+                    }
+                }else{
```
Comment (sghpjuikit) [inline modules/javafx.controls/src/main/java/javafx/scene/control/skin/TableRowSkinBase.java]: `?else` formatting

File: modules/javafx.controls/src/main/java/javafx/scene/control/skin/TableRowSkinBase.java [hunk 3]
```diff
@@ -339,45 +339,79 @@ public TableRowSkinBase(C control) {
...
-            if (isVisible) {
-                if (fixedCellSizeEnabled && tableCell.getParent() == null) {
-                    getChildren().add(tableCell);
+            final boolean visible = isOverlap(start, end, scrollX, headerWidth + scrollX);
+            if(visible) {
+                if(firstVisibleColumnIndex == -1) {
+                    firstVisibleColumnIndex = column;
                 }
+                lastVisibleColumnIndex = column;
+            }else if( firstVisibleColumnIndex != -1 ) {
```
Comment (sghpjuikit) [inline modules/javafx.controls/src/main/java/javafx/scene/control/skin/TableRowSkinBase.java]: `}else` formatting

File: modules/javafx.controls/src/main/java/javafx/scene/control/skin/TableRowSkinBase.java [hunk 4]
```diff
@@ -339,45 +339,79 @@ public TableRowSkinBase(C control) {
         int index = control.getIndex();
         if (index < 0/* || row >= itemsProperty().get().size()*/) return;
 
+        int firstVisibleColumnIndex = -1;
+        int lastVisibleColumnIndex = -1;
+        final VirtualFlow<?> virtualFlow = getVirtualFlow();
+        final double scrollX = virtualFlow == null ? 0.0 : virtualFlow.getHbar().getValue();
+        final Insets padding = getSkinnable().getPadding();
+        final double vfWidth = virtualFlow == null ? 0.0:virtualFlow.getWidth();
```
Comment (sghpjuikit) [inline modules/javafx.controls/src/main/java/javafx/scene/control/skin/TableRowSkinBase.java]: `0.0:` formatting

File: modules/javafx.controls/src/main/java/javafx/scene/control/skin/TableRowSkinBase.java [hunk 5]
```diff
@@ -339,45 +339,79 @@ public TableRowSkinBase(C control) {
...
+        double layoutY = snappedTopInset();
+        final double snapSizeYVerticalPadding = snapSizeY(verticalPadding);
+
+        for (int column = 0, max = cells.size(); column < max; column++) {
+            R tableCell = cells.get(column);
+            TableColumnBase<T, ?> tableColumn = getTableColumn(tableCell);
 
+            boolean isVisible = firstVisibleColumnIndex <= column && column <= lastVisibleColumnIndex;
+            width = snapSizeX(tableColumn.getWidth());
+            if (isVisible || isOverlap(tableCell.getLayoutX(), tableCell.getLayoutX()+width, scrollX, headerWidth + scrollX)) {
```
Comment (sghpjuikit) [inline modules/javafx.controls/src/main/java/javafx/scene/control/skin/TableRowSkinBase.java]: `)+w` formatting

**Review by sghpjuikit — COMMENTED** [review]
Hi @sghpjuikit, thanks for making a comment in an OpenJDK project!

All comments and discussions in the OpenJDK Community must be made available under the OpenJDK [Terms of Use](https://openjdk.java.net/legal/tou/). If you already are an OpenJDK [Author](https://openjdk.java.net/bylaws#author), [Committer](https://openjdk.java.net/bylaws#committer) or [Reviewer](https://openjdk.java.net/bylaws#reviewer), please click [here](https://bugs.openjdk.java.net/secure/CreateIssue.jspa?pid=11300&issuetype=1) to open a new issue so that we can record that fact. Please Use "Add GitHub user sghpjuikit for the summary.

If you are not an OpenJDK Author, Committer or Reviewer, simply check the box below to accept the OpenJDK Terms of Use for your comments.

 - [ ]  I agree to the [OpenJDK Terms of Use](https://openjdk.java.net/legal/tou/) for all comments I make in a project in the [OpenJDK GitHub organization](https://github.com/openjdk).

Your comment will be automatically restored once you have accepted the OpenJDK [Terms of Use](https://openjdk.java.net/legal/tou/).

**Comment by nlisker** [comment]
The title of your PR should match the title of the JBS issue. Moreover, #108 already tackles this issue with a different approach and 2 PRs on the same issue can't be intetgrated. Since I'm not familiar with this code, I can't advise on who's solution is more suitable. I suggest you discuss both of the approaches on the mailing list. I saw many JBS issues around this performance issue, probably one of them fits (or a new one can be created).

**Comment by robilad** [comment]
Hi,
I couldn't find you listed at https://www.oracle.com/technetwork/community/oca-486395.html . Please send me an e-mail at dalibor.topic@oracle.com with information about your OCA, so that I can look it up.

**Comment by yososs** [comment]
My name is "Naohiro Yoshimoto" 
I have received an approved email on 2019-8-3.

**Comment by yososs** [comment]
19fabf2eafcb02b519d39a1b0a9dad5b9209db64 

* The constructor creates a cell, but the
 If the fixedCellSize is not initialized, the
Fixed a wasteful process of creating all cells and deleting cells out of the display range.
* Reduce the load on ExpressionHelper.removeListener by removing cells outside of the display range from the back of the scrolling operation.

**Comment by yososs** [comment]
When setting the cell height to a fixed value with setFixedCellSize(), column virtualization can improve performance.

As the next step, I think it is possible to try to enable column virtualization regardless of whether setFixedCellSize() is set or not.
Before doing this step, the direct access validation of the internal structure of the test code needs to be changed via the public API.

**Comment by kleopatra** [comment]
this indeed improves scrolling performance considerably (from extremely lagging to smooth following the mouse when dragging the thumb of the vertical scrollbar). 

As a side-effect, start-up time of TreeTableView seems to increase considerably (at least by a factor of 3 to 4, probably not linear in # of columns) - see f.i. the [example in JDK-8166956](https://bugs.openjdk.java.net/browse/JDK-8166956). Any idea why that might happen?

**Comment by yososs** [comment]
Since there was a problem that the cell was not updated when the window was maximized, I added a fix to TreeTableViewSkin.java.

Added test code (BigTreeTableViewTest) for TreeTableView to the first post.

@kleopatra 
Does the test code I added (BigTreeTableViewTest.java) also have side effects?

**Comment by kleopatra** [comment]
> 
> 
> Since there was a problem that the cell was not updated when the window was maximized, I added a fix to TreeTableViewSkin.java.
> 

that's just the added listener to hbar properties (same as in TableViewSkin), or anything else changed? 

> Added test code (BigTreeTableViewTest) for TreeTableView to the first post.
> 
> @kleopatra
> Does the test code I added (BigTreeTableViewTest.java) also have side effects?

yeah, but why do you ask me - you could easily see for yourself :) And don't understand why you wrap the hierarchy setup in runlater (which smears a bit over the delay by filling the treeTable content at an unspecific time "later")

**Comment by yososs** [comment]
When the startup time is measured by eye, the impression changes depending on the individual difference.
The effect of runLater affects your experience.

However, I succeeded in further improving performance by eliminating another bottleneck in TreeTableView. Of course, it also includes improvements in startup time.

I plan to commit at a later date.

**Comment by kleopatra** [comment]
> 
> 
> When the startup time is measured by eye, the impression changes depending on the individual difference.

my eye is a precision instrument :) Seriously, who would do such a thingy? Obviously, it must be measured, for zeroth approximation doing so crudely by comparing currentMillis before/after showing is good enough to "see" the tendency.

> The effect of runLater affects your experience.

that's why you shouldn't do it when trying to gain insight into timing issues ;)

> 
> However, I succeeded in further improving performance by eliminating another bottleneck in TreeTableView. Of course, it also includes improvements in startup time.

cool :)

**Comment by yososs** [comment]
Column virtualization causes shortness of breath when scrolling a large stroke horizontally. 
This does not happen when stroking on the scrollbar. Looks like a potential problem with VirtualFlow.

If you are worried about shortness of breath, the following code will help reduce the problem.


``` Java
　private static final int OVERLAP_MARGIN = 500;

    private static boolean isOverlap(double start, double end, double start2, double end2){
    	start = Math.max(start-OVERLAP_MARGIN, start2);
    	end = Math.min(end+OVERLAP_MARGIN, end2);
        return (start<=end2 && end >= start2);
    }
```

**Comment by kevinrushforth** [comment]
@yososs Per [this message](https://mail.openjdk.java.net/pipermail/openjfx-dev/2020-September/027534.html) on the openjfx-dev mailing list, I have closed [JDK-8185886](https://bugs.openjdk.java.net/browse/JDK-8185886) as a duplicate, and suggested another existing JBS issue for this PR to use. Please change the title to:

```
8185887: TableRowSkinBase fails to correctly virtualize cells in horizontal direction
```

**Comment by yososs** [comment]
Below are the answers to JBS's suggestions.

> The patch below resolves the issue, but it is not likely the correct solution (we are trying to determine the table width, but we are getting the width and padding on the underlying virtualflow (the width is fine, the padding should really come from tableview directly): 

You probably mention this part,
At least padding refers to Skinnable(TableView or TreeTableView). This is the same as before (L683). The width of VirtualFlow is determined by the header of Table.

https://github.com/openjdk/jfx/blob/c81dd2e2fe935fa87c3e678cdadeb2f977f5975f/modules/javafx.controls/src/main/java/javafx/scene/control/skin/TableRowSkinBase.java#L360-L365

For example, can you assume a case that does not work properly?

**Comment by yososs** [comment]
This is the answer to JBS's comment.

The BigTreeTableViewTest.java attached to this PR commentary is a modified version of the JDK-8166956 test code. The original test code is good for reproducing the problem, but I have decided that it cannot be used as a test code because the cell values are random and the cell values change randomly with the close / expand operation.

**Comment by sghpjuikit** [comment]
I'm writing this from years long frustration with the way JavaFX has been and still is being handled.

It is disheartening to see PR's like this to go stale. I see this happen again again in JavaFX - any bigger PR, particularly those related to VirtualFlow, virtualized controls and listeners go stale. Every single time. I thought someone should say a word or two.

It is good to have a certain quality standard for the code from contributors, but I'm noticing a pattern, where there is either an onslaught of demotivating responses from maintainers towards PR owners, usually subtle, but demotivating as an (in my opinion) ridiculous amount of demand for effort is pertained in them. Or there is no help from the maintainers, causing the owner to lose interest, probably due to time constraints and having blew too much effort on the issue with no hope it going anywhere. This are the usual suspects:
- write complete tests for each little thing that can be useful
- this is good but it can be better - rework it
- you did not do enough research
- go sign bunch of weird stuff so we do not ignore you

Why can not the maintainers join efforts with the contributors, lead them towards successful merge and help grow the community? And finish valuable PRs? Instead contributors are expected to provide 100%, or even 120% solution or go home. This alienates like 99% of contribution. This is not contributor friendly and a sole reason I have never contributed to this project (and believe me I would probably be the most active contributor!). I do not understand the OCA thingy at all (and i will never submit my name or organization name to a company so I can contribute to an open source project). I do not understand why there is time to rename bunch of useless code to make the inclusive bunch happy, but not enough time to fix 'remove throw UnsupportedException' bug people reported 3 years ago, when it takes 10 seconds to fix. You got the priorities straight alright. It seems to me as if there is no interest on part of maintainers to see the PR's actually get merged. It is sad, very sad. Every PR older than few weeks is a testament to my words. The contributors do not follow up their PR's for a reason, did anyone in the jfx project ever think of that?

I find it extremely funny (and disturbing at the same time) that you are even bot-removing my suggestions and comments in PR's now, again because I haven't signed some cursed thing. What are you a GDPR company? I will NOT sign it, now stop wasting my time with notifications.
Oracle is literally removing my comments on an external site. I consider these actions those of a censorship. But whatever.

It seems like JavaFX is a cursed project, from inception. Script, then Java. Standalone, then JDK. Then opensource, but anti-contributor. Jars, then modules. Painful setup. Broken Gradle plugin. Forever private APIs. Simple bugs, unfixed for years. Et cetera. I think most of this curse is from the stiff corporate policies, that wont let it catch a breath (not that anyone cares about my opinion on the subject). The only people that can give needed life into the project are contributors and you make it very hard for them to feel motivated. This project needs massive contributions to stay alive (it is not even feature complete, requiring swing here and there). It is your job to enable people to do those contributions. Do something before even your most rabid fans leave. I just did.

TLDR: After having comments removed for not signing up to Oracle corporate policies, a fan, potential contributor and long time bug reporter decides to stop contributing, commenting PR's and reporting bugs.

PS: I don't care if this comment belongs here or not.
 
/rant, have a nice day

File: modules/javafx.controls/src/main/java/javafx/scene/control/skin/TableRowSkinBase.java (hunk 1 above)
Comment (sghpjuikit) [reply modules/javafx.controls/src/main/java/javafx/scene/control/skin/TableRowSkinBase.java:501]: no

File: modules/javafx.controls/src/main/java/javafx/scene/control/skin/TableRowSkinBase.java (hunk 2 above)
Comment (sghpjuikit) [reply modules/javafx.controls/src/main/java/javafx/scene/control/skin/TableRowSkinBase.java:392]: no

File: modules/javafx.controls/src/main/java/javafx/scene/control/skin/TableRowSkinBase.java (hunk 3 above)
Comment (sghpjuikit) [reply modules/javafx.controls/src/main/java/javafx/scene/control/skin/TableRowSkinBase.java:376]: no

File: modules/javafx.controls/src/main/java/javafx/scene/control/skin/TableRowSkinBase.java (hunk 4 above)
Comment (sghpjuikit) [reply modules/javafx.controls/src/main/java/javafx/scene/control/skin/TableRowSkinBase.java:363]: no

File: modules/javafx.controls/src/main/java/javafx/scene/control/skin/TableRowSkinBase.java (hunk 5 above)
Comment (sghpjuikit) [reply modules/javafx.controls/src/main/java/javafx/scene/control/skin/TableRowSkinBase.java:423]: no
